### ISSUE TITLE
Consul class recommended/intended usage 

### ISSUE THREAD

**Issue by miseri** [oliora/ppconsul#46]
I haven't looked into this deeper, but when using the same Consul object with multiple Kv objects I ran into situations where curl segfaulted consistently.
After changing the code to use a new Consul objects per blocking key-value read, the segfaults went away. 

Just leaving this here in case someone else runs into this. Perhaps a note in the documentation on recommended usage of the Consul, Kv classes would be helpful?

Thanks for the library. I've only been looking at it for a couple of days but it has been useful so far. The vcpkg integration made it very easy to integrate.

**Comment by oliora** [comment]
Thank you for reporting this. Indeed I don’t have any notes on thread safety and recommended usage of Ppconsul from multiple threads.
Your conclusion is correct: `Consul` object can’t be used from multiple threads in the same time: one is purely single threaded so if you need to run parallel requests (like a blocking KV query) from multiple threads you need multiple `Consul` objects.

**Comment by miseri** [comment]
Thanks for your quick response. Added a PR to add your comments to the README

**Comment by oliora** [comment]
Merged. Thank you for the PR. I think I’ll change the threading model for Ppconsul soon: the Consul object will be more like a thread safe pool of connections so it can be called from multiple threads. Endpoint objects (`Kv` and so on) are already immutable thus thread safe.

**Comment by jgornowich** [comment]
I'd be interested in your ideas for a new threading model.  Indeed where I have used this currently in a multi thread environment I have separate Consul objects for each query thread. 

When you say the endpoint objects are already immutable, you are just meaning because they are json11::Json objects?

**Comment by oliora** [comment]
TLDR everything in Ppconsul will become thread safe (can be called from multple threads in the same time). Longer answer below.

By “endpoint” classes (bad name perhaps) I meant Ppconsul classes like `Kv`, `Agent` etc. They have no mutable state and all their methods can be marked as const.

With the new model you need only one `Consul` class instance for all threads (as far as it communicates with the same Consul server) and one instance of each “endpoint” class (`Kv`, `Agent` etc).

I guess I’ll implement the new model in upcoming days or week so you can see it

**Comment by oliora** [comment]
I've refactored Consul to make it thread-safe https://github.com/oliora/ppconsul/commit/2bc79c7401f43f6b93a815d205c207c041b8267b
